kernel/src/log_compaction/writer.rs

```diff
@@ -0,0 +1,177 @@
+        // Validate that the requested version range is within the snapshot's range
+        let snapshot_end_version = self.snapshot.version();
+        if self.end_version > snapshot_end_version {
+            return Err(Error::generic(format!(
+                "End version {} exceeds snapshot version {}",
+                self.end_version, snapshot_end_version
+            )));
+        }
+
+        // Create a log segment specifically for the compaction range
+        // This ensures we only process commits in [start_version, end_version]
+        let compaction_log_segment = LogSegment::for_table_changes(
+            engine.storage_handler().as_ref(),
+            self.snapshot.log_segment().log_root.clone(),
+            self.start_version,
+            Some(self.end_version),
+        )?;
+
+        // Read actions from the version-filtered log segment
+        let actions_iter = compaction_log_segment.read_actions(
+            engine,
+            COMPACTION_ACTIONS_SCHEMA.clone(),
+            COMPACTION_ACTIONS_SCHEMA.clone(),
+            None, // No predicate - we want all actions in the version range
+        )?;
+
+        let min_file_retention_timestamp_millis = self.deleted_file_retention_timestamp()?;
+
+        // Create checkpoint log replay processor for compaction
+        // This reuses the same reconciliation logic as checkpoints
+        let processor = CheckpointLogReplayProcessor::new(
+            min_file_retention_timestamp_millis,
+            self.get_transaction_expiration_timestamp()?,
+        );
+
+        // Process actions using the same iterator pattern as checkpoints
+        // The processor handles reverse chronological processing internally
+        let result_iter = processor.process_actions_iter(actions_iter);
+
+        // Wrap the iterator in a LogCompactionDataIterator to track action counts lazily
+        Ok(LogCompactionDataIterator {
+            compaction_batch_iterator: Box::new(result_iter),
+            actions_count: 0,
+            add_actions_count: 0,
+        })
+    }
+}
+
+/// Iterator over log compaction data. Provides the reconciled actions that should be written
+/// to the compaction file.
+pub struct LogCompactionDataIterator {
+    /// The nested iterator that yields compaction batches with action counts
+    pub(crate) compaction_batch_iterator:
+        Box<dyn Iterator<Item = DeltaResult<CheckpointBatch>> + Send>,
+    /// Running total of actions included in the compaction
+    pub(crate) actions_count: i64,
+    /// Running total of add actions included in the compaction
+    pub(crate) add_actions_count: i64,
+}
+
+impl LogCompactionDataIterator {
+    /// Get the total number of actions in the compaction
+    pub fn total_actions(&self) -> i64 {
+        self.actions_count
+    }
+
+    /// Get the total number of add actions in the compaction
+    pub fn total_add_actions(&self) -> i64 {
+        self.add_actions_count
+    }
+}
```

Comment:
in checkpoints IIRC we use this to write out the `_last_checkpoint` file - while these do seem somewhat useful to know, do we have an immediate need for them? (perhaps they will become useful for stats in the future but if stats are primary consumer perhaps we keep `pub(crate)` until we decide how to expose them?)

Reply:
We will need to remove the #[allow(unused)] when we add integration tests.